[VarDumper] Add `#[SensitiveElement]` attribute

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | _NA_
| License       | MIT
| Doc PR        | Todo

Following the same dynamic as PHP 8.2 `#[\SensitiveParameter]`, an attribute to tell the cloners not to clone a set of classes/properties would be nice. Thanks to a new `#[SensitiveElement]` attribute, this allows us to define what we don't want to be cloned because of the sensitive nature of data. Instead, properties and classes marked as sensitive are replace by a new `SensitiveElementStub`. This allows, among other things, to hide the content of these elements when dumping their content with VarDumper. Here is the attribute in action:

```php
namespace App\Controller;

use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\VarDumper\Attribute\SensitiveElement;

// ...

#[SensitiveElement]
class Foo
{
}

class Qux
{
    #[SensitiveElement]
    private string $ownPassword = "something else";

    private string $publicInfo = 'Public one';
}

class Bar
{
    public Foo $foo;

    public Route $qux;

    public Qux $qux2;

    #[SensitiveElement]
    private string $password = 'root';

    public function __construct()
    {
        $this->foo = new Foo();
        $this->qux = new Route();
        $this->qux2 = new Qux();
    }
}
```

Dumping with VarDumper an instance of `Bar` results on this:

<img width="559" alt="Capture d’écran 2023-02-01 à 22 07 48" src="https://user-images.githubusercontent.com/2144837/216163576-2e824df9-982c-47e1-92dd-1a55e2d863f0.png">

alexandre-daubois committed:
[VarDumper] Add `#[SensitiveElement]` attribute

nicolas-grekas commented:
I like the idea of using attributes to fine tune how a class is dumped.
But the current implementation is going to be a perf hog, creating reflection objects at a hit rate.

I would like we have one call to reflection per class, and there is already one, cached, in `AbstractCloner::castObject()`.
I'd suggest fetching+caching attributes there.
I'd also suggest to look only at the class level. For property-specific attributes, we should add a "property" parameter to a class attribute:
```
#[SensitiveValue(property: 'foo')]
class Bar
{
    private $foo;
}
```

I would also suggest to make all VarDumper's attribute extend an base class. That'd allow getting all attributes in one call:
`$r->getAttributes(TheBaseClass::class, ReflectionAttribute:IS_INSTANCEOF)`

Then, I would look at existing casters and see if we couldn't replace some of them by attributes, eg `StubCaster::cutInternals`, or make it possible to declare properties that should use EnumStub inline, things like that.

nicolas-grekas commented:
Note also that we'd need a flag to ignore attributes if one wants to. Eg I might not be interested in hidding sensitive values, when it conflicts with my debugging experience.

alexandre-daubois commented:
Thank you very much for this wonderful feedback Nicolas!

I updated the PR. The reflection is only done once now and is cached with the rest of class' info. I also added a flag field+constant in `AbstractCloner`. Depending on which PR comes first, I would definitely add a `_dumpSensitive` special argument to `dump()`/`dd()` in https://github.com/symfony/symfony/pull/48667.

There's a new `AbstractAttribute` in VarDumper's namespace. Not exactly sure about the naming yet. But it's here and used by `AbstractCloner`. `SensitiveElement` attribute now accepts one or more properties. This solution suits me fine, and seems to ease the code a bit too!

I'll have a look at casters that might be replaced by attributes. Keeping your comment's link in my notes 👍

alexandre-daubois commented:
Fixed a test case + added properties info also in the cache of the cloner. This will later allow to easily parse properties attributes to create attributes like `DumpAsEnumStub` for instance, as pointed by Nicolas.

nicolas-grekas commented on src/Symfony/Component/VarDumper/Attribute/AbstractAttribute.php:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ *
+ * @internal
```
it shouldn't be internal to be useful for other use cases

nicolas-grekas commented on src/Symfony/Component/VarDumper/Attribute/SensitiveElement.php:
```diff
@@ -0,0 +1,33 @@
+#[\Attribute(\Attribute::TARGET_CLASS)]
+class SensitiveElement extends AbstractAttribute
+{
+    private array $properties;
```
should be public + no getter

nicolas-grekas commented on src/Symfony/Component/VarDumper/Cloner/AbstractCloner.php:
```diff
@@ -336,17 +350,48 @@ protected function castObject(Stub $stub, bool $isNested): array
+            $classAttributes = $r->getAttributes(AbstractAttribute::class, \ReflectionAttribute::IS_INSTANCEOF);
+
+            $properties = [];
+            foreach ($r->getProperties() as $property) {
```
no way we call getProperties in cloners, this will kill perf + the code is broken because it doesn't get private properties on parent classes

nicolas-grekas commented on src/Symfony/Component/VarDumper/Cloner/AbstractCloner.php:
```diff
@@ -336,17 +350,48 @@ protected function castObject(Stub $stub, bool $isNested): array
+        foreach ($classAttributes as $attribute) {
+            $attribute = $attribute->newInstance();
+
+            if ($attribute instanceof SensitiveElement && !($this->flags & self::CLONER_WITH_SENSITIVE_ELEMENTS)) {
```
this whole concern should be dealt with in a caster to me - in Caster::castObject here
we'd need to find a way to pass the attributes to them

nicolas-grekas commented on src/Symfony/Component/VarDumper/Cloner/Data.php:
```diff
@@ -351,6 +352,12 @@ private function dumpItem(DumperInterface $dumper, Cursor $cursor, array &$refs,
                     // no break
                 case Stub::TYPE_OBJECT:
+                    if ($item->sensitive) {
+                        $this->dumpItem($dumper, $cursor, $refs, new SensitiveElementStub($item->class));
```
that's for casters

nicolas-grekas commented on src/Symfony/Component/VarDumper/Cloner/Visibility.php:
```diff
@@ -0,0 +1,24 @@
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+enum Visibility
```
enums are useless, one more example :P

nicolas-grekas reviewed (commented):
On second thought, I'm not really sold on the need for SensitiveElement. When you dump, you want to see the state :shrug: 
But I'm curious about what we can do with attributes and how we can use them to configure casters.
I also get this might be too much work for something that doesn't have a clear use case and is an experiment with no definitive promise of achieving something really useful, so feel free to close.

alexandre-daubois commented:
Thanks for your feedback @nicolas-grekas! I understand. I started to experiment a few things with casters and attributes. I'll see where it goes.
Thanks for the time spent on the review!
